// heroku-cli-command/lib/mixins/app.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _get2 = require('babel-runtime/helpers/get');

var _get3 = _interopRequireDefault(_get2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

exports.default = function (Command) {
  var _class, _temp;

  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  return _temp = _class = function (_Command) {
    (0, _inherits3.default)(AppMixin, _Command);

    function AppMixin() {
      (0, _classCallCheck3.default)(this, AppMixin);
      return (0, _possibleConstructorReturn3.default)(this, (AppMixin.__proto__ || (0, _getPrototypeOf2.default)(AppMixin)).apply(this, arguments));
    }

    (0, _createClass3.default)(AppMixin, [{
      key: 'init',
      value: function () {
        var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee() {
          return _regenerator2.default.wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  _context.next = 2;
                  return (0, _get3.default)(AppMixin.prototype.__proto__ || (0, _getPrototypeOf2.default)(AppMixin.prototype), 'init', this).call(this);

                case 2:
                  if (!(!this.app && options.required !== false)) {
                    _context.next = 4;
                    break;
                  }

                  throw new Error('No app specified');

                case 4:
                case 'end':
                  return _context.stop();
              }
            }
          }, _callee, this);
        }));

        function init() {
          return _ref.apply(this, arguments);
        }

        return init;
      }()
    }, {
      key: 'app',
      get: function get() {
        if (this._app) return this._app;
        if (typeof this.flags.app === 'string') this._app = this.flags.app;
        return this._app;
      }
    }]);
    return AppMixin;
  }(Command), _class.flags = [APP_FLAG], _temp;
};

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var APP_FLAG = {
  name: 'app',
  char: 'a',
  description: 'app to run command against',
  hasValue: true
};
/* globals
   Class
 */
