// heroku-cli-command/lib/mixins/heroku.js
'use strict';

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _get2 = require('babel-runtime/helpers/get');

var _get3 = _interopRequireDefault(_get2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var http = require('../http');

module.exports = function () {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  return function (superclass) {
    return function (_superclass) {
      (0, _inherits3.default)(_class, _superclass);

      function _class() {
        (0, _classCallCheck3.default)(this, _class);
        return (0, _possibleConstructorReturn3.default)(this, (_class.__proto__ || (0, _getPrototypeOf2.default)(_class)).apply(this, arguments));
      }

      (0, _createClass3.default)(_class, [{
        key: 'init',
        value: function () {
          var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee() {
            var auth, netrc, host, HTTP;
            return _regenerator2.default.wrap(function _callee$(_context) {
              while (1) {
                switch (_context.prev = _context.next) {
                  case 0:
                    _context.next = 2;
                    return (0, _get3.default)(_class.prototype.__proto__ || (0, _getPrototypeOf2.default)(_class.prototype), 'init', this).call(this);

                  case 2:
                    auth = process.env.HEROKU_API_KEY;

                    if (!auth) {
                      netrc = require('netrc')();
                      host = netrc['api.heroku.com'];

                      if (host) auth = host.password;
                    }

                    if (!(!auth && options.required !== false)) {
                      _context.next = 6;
                      break;
                    }

                    throw new Error('Not logged in');

                  case 6:
                    HTTP = require('http-call');

                    this.heroku = new HTTP({
                      host: 'api.heroku.com',
                      protocol: 'https:',
                      requestMiddleware: http.requestMiddleware.bind(this),
                      responseMiddleware: http.responseMiddleware.bind(this),
                      auth: ':' + auth,
                      headers: {
                        'user-agent': 'heroku-cli/' + this.options.version,
                        'accept': 'application/vnd.heroku+json; version=3'
                      }
                    });

                  case 8:
                  case 'end':
                    return _context.stop();
                }
              }
            }, _callee, this);
          }));

          function init() {
            return _ref.apply(this, arguments);
          }

          return init;
        }()
      }]);
      return _class;
    }(superclass);
  };
};
