// heroku-cli-command/lib/output/action.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Action = function () {
  function Action(out) {
    (0, _classCallCheck3.default)(this, Action);
    this.task = null;

    this.out = out;
  }

  (0, _createClass3.default)(Action, [{
    key: 'start',
    value: function start(message) {
      var status = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';

      var msg = message + '...';
      if (this.task) {
        if (this.task.spinner) {
          this.task.spinner.text = msg;
          this.task.spinner.status = status;
        } else process.stderr.write('\n' + msg + ' ' + status);
      } else {
        this.task = { message: message, status: status };
        if (this.displaySpinner) {
          var _Spinner = require('./spinner');
          this.task.spinner = new _Spinner({ text: msg, command: this.out, status: status });
          this.task.spinner.start();
        } else this.out.stderr.write(msg + ' ' + status);
      }
    }
  }, {
    key: 'stop',
    value: function stop() {
      var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'done';

      if (!this.task) return;
      this.start(this.task.message, msg);
      if (this.task && this.task.spinner) this.task.spinner.stop(msg);else process.stderr.write('\n');
      delete this.task;
    }
  }, {
    key: 'pause',
    value: function pause(fn) {
      var spinner = this.task ? this.task.spinner : null;
      if (spinner) {
        if (spinner) {
          spinner.stop();
          spinner.clear();
        } else {
          delete this.task;
          process.stderr.write('\n');
        }
      }
      fn();
      if (spinner) {
        process.stderr.write('\n');
        spinner.start();
      }
    }
  }]);
  return Action;
}();

exports.default = Action;
